### PR TITLE
Lint: Lint rule for use of UIUtils.showSnackbar instead of Snackbar.make

### PR DESCRIPTION
**Pull request by mrudultora**
## Purpose / Description
The methods to use Snackbars must be from UIUtils class rather than using them directly (Snackbar.make(...)).

## Fixes
None

## Approach
Added a lint rule to check this. And, removed the currently used Snackbar.make(....).

Untitled

# Screenshots
img

This screenshot is fix for the existing lint errors (related to this PR).

## How Has This Been Tested?
Unit tests passing and `gradlew lintRelease` successful.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 1]
```diff
@@ -3815,15 +3815,14 @@ private void showDeveloperContact(int errorCode) {
...
-        Snackbar snackbar = Snackbar.make(parentLayout, snackbarMsg, Snackbar.LENGTH_LONG);
-        View snackbarView = snackbar.getView();
-        TextView snackTextView = snackbarView.findViewById(com.google.android.material.R.id.snackbar_text);
-        snackTextView.setTextColor(Color.WHITE);
-        snackTextView.setMaxLines(3);
-
-        snackbar.setActionTextColor(Color.MAGENTA)
-                .setAction(getString(R.string.reviewer_invalid_api_version_visit_documentation), view -> openUrl(Uri.parse("https://github.com/ankidroid/Anki-Android/wiki")));
 
+        Snackbar snackbar = UIUtils.showSnackbar(this,
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3819]: Please do this in a separate commit before adding the lint

Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3819]: I don't think this is correct. This snackbar requires 3 lines, which the previous implementation did not provide

**Review by david-allison-1 — COMMENTED** [review]
Thanks!

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java (hunk 1 above)
Comment (mrudultora) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3819]: Would you suggest anything to `setMaxLines` to 3.

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3819]: No, just in this case. Would be best with an `@SuppressLint` on the method with an explanation that this is a special case

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 2]
```diff
@@ -3815,15 +3815,14 @@ private void showDeveloperContact(int errorCode) {
...
+                snackbarMsg,
+                false,
+                R.string.reviewer_invalid_api_version_visit_documentation,
+                view -> openUrl(Uri.parse("https://github.com/ankidroid/Anki-Android/wiki")),
+                parentLayout,
+                null);
         snackbar.show();
```
Comment (nicdard) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3826]: Also, in case we retain the `UIUtils.showSnackbar` call, we should get rid of this line.
Working out on this observation, I would suggest that maybe we should advice the developer to use `UIUtils.getSnackbar`, but I am not really sure about this.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java (hunk 1 above)
Comment (mrudultora) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3819]: @david-allison-1 I can still resolve this. Fetched the textview by (`snackbar.getView().findViewById`) and then setting `maxLines`. It's working fine.

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/DirectSnackbarMakeUsage.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.uast.UCallExpression;
+import org.jetbrains.uast.UClass;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This custom Lint rules will raise an error if a developer uses the {android.widget.Toast#makeText(...)} method instead
+ * of using the method provided by the UIUtils class {com.ichi2.anki.UIUtils#showThemedToast(...)}.
```
Comment (nicdard) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/DirectSnackbarMakeUsage.java:41]: The comment here should reference Snackbars :)

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java (hunk 2 above)
Comment (mrudultora) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3826]: @nicdard I think we must not use `getSnackbar` directly.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java (hunk 1 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3819]: If it works, I'm happy - if you could test it, it'd be fantastic, but if it's tricky then no bother

Comment (mrudultora) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3819]: Yep, it's working. I will attach a screenshot showing the same.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 4]
```diff
@@ -3815,15 +3815,16 @@ private void showDeveloperContact(int errorCode) {
...
-                .setAction(getString(R.string.reviewer_invalid_api_version_visit_documentation), view -> openUrl(Uri.parse("https://github.com/ankidroid/Anki-Android/wiki")));
 
+        Snackbar snackbar = UIUtils.showSnackbar(this,
+                snackbarMsg,
+                false,
+                R.string.reviewer_invalid_api_version_visit_documentation,
+                view -> openUrl(Uri.parse("https://github.com/ankidroid/Anki-Android/wiki")),
+                parentLayout,
+                null);
+        TextView snackbarTextView=snackbar.getView().findViewById(com.google.android.material.R.id.snackbar_text);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:3826]:
```suggestion
        TextView snackbarTextView = snackbar.getView().findViewById(com.google.android.material.R.id.snackbar_text);
```

**Review by david-allison-1 — COMMENTED** [review]
nitpick

**Comment by mrudultora** [comment]
@david-allison-1 Changes done. Also added the screenshot in the PR template.

**Review by david-allison-1 — APPROVED** [review]
LGTM

**Review by mikehardy — APPROVED** [review]
LGTM - will squash so the lint doesn't fail before the fix commit

**Comment by mikehardy** [comment]
Hi there!

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for April 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in April)

Thanks!
